core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -684,11 +690,15 @@ class ReplicaManager(val config: KafkaConfig,
         else
           getReplicaOrException(tp)
 
-        // decide whether to only fetch committed data (i.e. messages below high watermark)
-        val maxOffsetOpt = if (isolationLevel == IsolationLevel.READ_COMMITTED)
+        val initialHighWatermark = localReplica.highWatermark.messageOffset
+        val lastStableOffset = if (isolationLevel == IsolationLevel.READ_COMMITTED)
           Some(localReplica.lastStableOffset.messageOffset)
-        else if (readOnlyCommitted)
-          Some(localReplica.highWatermark.messageOffset)
+        else
+          None
+
+        // decide whether to only fetch committed data (i.e. messages below high watermark)
+        val maxOffsetOpt = if (readOnlyCommitted)
```

Comment:
It's a bit confusing that we now have 2 slightly different meanings for `committed`.